File: src/recent-reads/gen.py
import re
import sys
import html

def process(txt):
    txt = html.escape(txt)
    for src, dst in [
        (r'(?<!\\)\*(.*?)(?<!\\)\*', r'<em>\1</em>'),
        (r'(?<!\\)`(.*?)(?<!\\)`', r'<code>\1</code>'),
        (r'\[(.*?)\]\((.*?)\)', r'<a href="\2">\1</a>'),
        (r'\\([*`])', r'\1'),
    ]:
        txt = re.sub(src, dst, txt)
    return txt

print('''
{% extends layout %}

{% block head %}
  <title>Recent Readings - Kevin Cao</title>
  <meta name="description" content="" />
{% endblock %}

{% block content %}
  <h1 class="text-center">Recent Reads</h1>
  <p>I sometimes stumble upon really unique websites on the Internet that I wish to share. Also some interesting articles and videos.</p> 
''')

print('<section>')
with open('/home/kjc/recent-reads', 'r') as fp:
    cards = []
    for l in fp.readlines():
        date, url, desc = l.split(maxsplit=2)
        if date == 'NSFW':
            date, url, desc = l.split(maxsplit=3)[1:]
            continue
        type, desc = desc.split(':', maxsplit=1)
        cards.append(f'<div class="inline-block mb-4"><a title="added on {date}" href="{url}"><b>{type}</b></a> {process(desc)}</div>')
    for i in range(3):
        print('<div class="p-2 sm:float-left sm:w-[calc(33.33%-1rem)]">')
        for c in cards[i::3]:
            print(c)
        print('</div>')
print('</section>')

print('{% endblock %})')

File: src/fun-websites/gen.jl
# println("""
# {% extends layout %}

# {% block head %}
#   <title>Recent Readings - Kevin Cao</title>
#   <meta name="description" content="" />
# {% endblock %}

# {% block content %}
#   <h1 class="center">Recent Reads</h1>
#   <p>A list of recent things on the Internet I have read/skimmed/watched and found entertaining.</p> 
# """)

# println("<ul>")
# open("/home/kjc/recent-reads", "r") do fp
#     while !eof(fp)
#         date, url, desc = split(readline(fp), limit=3)
#         println("  <li><a title=\"added on $date\" href=\"$url\">$desc</a></li>")
#     end
# end
# println("</ul>")

# println("{% endblock %})")

println("""
{% extends layout %}

{% block head %}
  <title>Recent Readings - Kevin Cao</title>
  <meta name="description" content="" />
  {% style %}
    pre {
      overflow-x: visible;
      white-space: pre-wrap;
      margin: 0;
    }
  {% endstyle %}
{% endblock %}

{% block content %}
  <h1 class="center">Recent Reads</h1>
  <p>A list of recent things on the Internet I have read/skimmed/watched and found entertaining.</p> 
""")

println("<pre><code>")
open("/home/kjc/recent-reads", "r") do fp
    while !eof(fp)
        date, url, desc = split(readline(fp), limit=3)
        println("<a title=\"added on $date\" href=\"$url\">link</a> $desc")
    end
end
println("</code></pre>")

println("{% endblock %})")
